File: app/src/main/java/com/snt/phoney/di/module/WxApiServiceModule.java
package com.snt.phoney.di.module;

import com.google.gson.Gson;
import com.snt.phoney.BuildConfig;
import com.snt.phoney.api.WxApi;
import com.snt.phoney.utils.adapter.LiveDataCallAdapterFactory;
import com.snt.phoney.utils.data.Constants;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class WxApiServiceModule {

    @Singleton
    @Provides
    @Named("wx_api")
    public static Retrofit provideWeiboApiRetrofit(Gson gson) {
        return new Retrofit.Builder()
                .baseUrl(Constants.Wechat.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addCallAdapterFactory(new LiveDataCallAdapterFactory())
                .client(getOkHttpClientBuilder().build())
                .build();
    }

    private static OkHttpClient.Builder getOkHttpClientBuilder() {
        OkHttpClient.Builder okHttpBuilder = new OkHttpClient.Builder();

        if (BuildConfig.DEBUG) {
            HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor();
            httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            okHttpBuilder.addInterceptor(httpLoggingInterceptor);
        }

        okHttpBuilder.addInterceptor(chain -> {
            Request request = chain.request();
            HttpUrl url = request.url().newBuilder()
                    .addQueryParameter("appid", Constants.Wechat.APP_ID)
                    .addQueryParameter("secret", Constants.Wechat.APP_SECRET)
                    .build();
            return chain.proceed(request.newBuilder().url(url).build());
        });
        return okHttpBuilder;
    }

    @Singleton
    @Provides
    public static WxApi provideWeiboApi(@Named("wx_api") Retrofit retrofit) {
        return retrofit.create(WxApi.class);
    }
}

File: app/src/main/java/com/snt/phoney/di/module/WxApiServiceModule.kt
package com.snt.phoney.di.module

import com.snt.phoney.BuildConfig
import com.snt.phoney.api.WeiboApi
import com.snt.phoney.api.WxApi
import com.snt.phoney.utils.adapter.LiveDataCallAdapterFactory
import com.snt.phoney.utils.adapter.ResponseConverterFactory
import com.snt.phoney.utils.data.Constants
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Named
import javax.inject.Singleton

@Module
object WxApiServiceModule {

    @JvmStatic
    @Singleton
    @Provides
    @Named("wx_api")
    fun provideWeiboApiRetrofit(): Retrofit {
        return Retrofit.Builder()
                .baseUrl(Constants.Wechat.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addCallAdapterFactory(LiveDataCallAdapterFactory())
                .client(getOkHttpClientBuilder().build())
                .build()
    }

    @JvmStatic
    private fun getOkHttpClientBuilder(): OkHttpClient.Builder {
        val okHttpBuilder = OkHttpClient.Builder()

        if (BuildConfig.DEBUG) {
            val httpLoggingInterceptor = HttpLoggingInterceptor()
            httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
            okHttpBuilder.addInterceptor(httpLoggingInterceptor)
        }

        okHttpBuilder.addInterceptor { chain ->
            val request = chain.request()
            val url = request.url().newBuilder()
                    .addQueryParameter("appid", Constants.Wechat.APP_ID)
                    .addQueryParameter("secret", Constants.Wechat.APP_SECRET)
                    .build()
            return@addInterceptor chain.proceed(request.newBuilder().url(url).build())
        }
        return okHttpBuilder
    }

    @JvmStatic
    @Singleton
    @Provides
    fun provideWeiboApi(@Named("wx_api") retrofit: Retrofit): WxApi = retrofit.create(WxApi::class.java)
}